server/scriptController.js

```
@@ -19,11 +21,110 @@ const scriptController = {
+    const scriptUpload = upload.single('newScript');
+
+    // actually upload the script
+    scriptUpload(req, res, next, (error) => {
+      if (error)
+        return next({
+          log: `error: ${error} occurred when uploading the script`,
+          message: 'error when uploading the script',
+        });
+      else {
+        return next();
+      }
+    });
+  },
+  parseScript: (req, res, next) => {
+    // run the script through playData and save the results to the database
+    const path = process.env.UPLOADPATH + req.file.filename;
+    const scriptData = parseScript(path);
+    playData[scriptData.title] = scriptData;
+    const characterObjects = Object.values(scriptData.characterObjs);
+
+    // check if the script already exists
+    db.query(`SELECT 'found' FROM scripts WHERE title=$1`, [scriptData.title])
+      .then((result) => {
+        if (result.rows.length > 0) {
+          throw new Error('Script title already exists');
+        }
+        // script is not in the database, add it
+        else {
+          return db.query('INSERT INTO scripts (title, filename) VALUES ($1, $2) RETURNING id', [
+            scriptData.title,
+            req.file.filename,
+          ]);
+        }
+      })
+      .then((result) => {
+        const id = result.rows[0].id;
+        return id;
+      })
+      .then((id) => {
+        // Build out the arrays of column values
+        const names = [];
+        const counts = [];
+        const speakNums = [];
+        for (let char of characterObjects) {
+          names.push(char.name);
+          counts.push(char.lineCount);
+          speakNums.push(char.speaksNum);
+        }
+        // add all the characters to the database
+        let text = `INSERT INTO characters (script_id, name, line_count, speaks_count) 
+          VALUES ($1, UNNEST($2::TEXT[]), UNNEST($3::INTEGER[]),UNNEST($4::INTEGER[]))`;
+
+        db.query(text, [id, names, counts, speakNums]).then(() => {
```

nit: .then should be on its own line like the others

---

probably due to db.query not being returned

---

this db.query may fail silently due to not having a .catch